### DISCUSSION TITLE
IR Sensors

### DISCUSSION THREAD

**Discussion by MikasaSucasa** [frc5567/2023-Robot#124] [Team Posts]
This is the code that adafruit gives for a demo. We can just add a string to the pixy code saying broken or unbroken. We will need someone to wire them up next time. I will solder them up if I find my other sensor.

**Comment by MikasaSucasa** [comment]
/* 
  IR Breakbeam sensor demo!
*/
 
#define LEDPIN 13
  // Pin 13: Arduino has an LED connected on pin 13
  // Pin 11: Teensy 2.0 has the LED on pin 11
  // Pin  6: Teensy++ 2.0 has the LED on pin 6
  // Pin 13: Teensy 3.0 has the LED on pin 13
 
#define SENSORPIN 4
 
// variables will change:
int sensorState = 0, lastState=0;         // variable for reading the pushbutton status
 
void setup() {
  // initialize the LED pin as an output:
  pinMode(LEDPIN, OUTPUT);      
  // initialize the sensor pin as an input:
  pinMode(SENSORPIN, INPUT);     
  digitalWrite(SENSORPIN, HIGH); // turn on the pullup
  
  Serial.begin(9600);
}
 
void loop(){
  // read the state of the pushbutton value:
  sensorState = digitalRead(SENSORPIN);
 
  // check if the sensor beam is broken
  // if it is, the sensorState is LOW:
  if (sensorState == LOW) {     
    // turn LED on:
    digitalWrite(LEDPIN, HIGH);  
  } 
  else {
    // turn LED off:
    digitalWrite(LEDPIN, LOW); 
  }
  
  if (sensorState && !lastState) {
    Serial.println("Unbroken");
  } 
  if (!sensorState && lastState) {
    Serial.println("Broken");
  }
  lastState = sensorState;
}

**Comment by coverbeek** [comment]
@JooshTheGreat and I were discussing when we both got home, and it seems that the break-beam is just a DIO input -- signaled or not. Josh posited that we don't even need the Arduino for this (and I tend to agree). We should be able to just wire directly to a DIO on the RIO and measure whether the circuit (beam) is closed/connected or not and react accordingly. Adding a field to the custom Arduino->RIO I2C data-struct created for the pixy comms may be instructive, but it seems like overkill since we need to react within the RIO code anyway -- the Arduino is just an extra step. The switch code Matt composed for the crate-stop switch would be exactly what we need here as well.

Thoughts? 

I'm fine with pursuing it using Arduino for instructive purposes, but it seems we may be better served showing how in this instance we're just transferring a simple on/off signal across an extra device if we use the Arduino.

**Comment by MikasaSucasa** [comment]
A pair is soldered together so it should be almost long enough to put on the robot thursday. Be thinking of how this can be implemented.

**Comment by MikasaSucasa** [comment]
That's fine Chris. I agree. However the coders wish to implement it, I will be happy either way. I posted the code because it is provided and how I tested and just re-verified that the system works.
I would assume that the RIO has a pullup resistor? That would be the only stipulation but if arduino has it, the rio should.

**Comment by coverbeek** [comment]
DIO on the RoboRIO has a pullup resistor for 5V, so we should be good. This is exactly how the double-pole switch works as well.

**Comment by coverbeek** [comment]
Maybe a good discussion to have Thursday - this decision is a preference, and we should talk though why it is so. In the end we can choose based on how maintainable the solution is, but both paths are equally valid and it would be good if everyone could understand why, at least at a high level.
